Source/Task/AsyncLib.cpp

```diff
@@ -425,32 +448,20 @@ static void CALLBACK CompletionCallbackForAsyncState(
     AsyncStateRef state;
     state.Attach(static_cast<AsyncState*>(context));
 
-    AsyncBlock* asyncBlock = state->providerData.async;
+    AsyncBlock* asyncBlock = state->userAsyncBlock;
     if (asyncBlock->callback != nullptr)
     {
         asyncBlock->callback(asyncBlock);
     }
 
```

Comment:
i wonder if you should pin this in a local var.  aka can asyncBlock->callback change after the if but before it exceutes?

Seems like 

auto callback = asyncBlock->callback;
if( callback != nullptr )
{
     callback(asyncBlock);
}

would be safer.  This applies here and else where in the file

Reply:
It would narrow the window but it's not legal to change the async block while the call is still active.